templates/edm.cpp.in

```diff
@@ -24,213 +24,40 @@
- * \a Args.
- */
-template<class EDMField, class PhotonEmitter, class ExchangeParticle>
-struct EDMVertexCorrectionSF {
-   static double value(const typename field_indices<EDMField>::type& indices,
-                       context_base& context);
-};
-
-/**
-* @class EDMVertexCorrectionFS
-* @brief A template that calculate contributions to the EDM
-*        of a given particle in a one loop diagram specified
-*        by a photon emitter and an exchange particle.
-* @tparam Args Specifies in order the field of which to
-*              calculate the electric dipole moment,
-*              the photon emitter and the exchange particle
-*              in a one-loop diagram where the photon emitter
-*              is a fermion and the exchange particle a scalar.
-*
-* This template evaluates the contribution to the electric
-* dipole moment of a one loop diagram with fields given by
-* \a Args.
-*/
-template<class EDMField, class PhotonEmitter, class ExchangeParticle>
-struct EDMVertexCorrectionFS {
-   static double value(const typename field_indices<EDMField>::type& indices,
-                       context_base& context);
-};
 } // anonymous namespace
 
 namespace flexiblesusy {
 namespace @ModelName@_edm {
-@EDM_InterfaceDefinitions@
-}
-} // namespace flexiblesusy
 
-namespace {
-/**
-* @defgroup LoopFunctions Loop functions
-* @brief The loop functions necessary for edm one-loop calculations.
-*
-* These are OneLoopFunctionA(), OneLoopFunctionB()
-* as specified in arXiv:0808.1819
-*/
+// spliting templated function into header and cpp
+// https://isocpp.org/wiki/faq/templates#separate-template-fn-defn-from-decl
+@calculateForwadDeclaration@
 
-double OneLoopFunctionA(double r)
+template <typename Lepton>
+double calculate_edm(const @ModelName@_mass_eigenstates& model, const softsusy::QedQcd& qedqcd@extraIdxDecl@)
 {
-   if (is_zero(r))
-      return std::numeric_limits<double>::infinity();
+   context_base context{ model };
 
-   const double d = r - 1.0;
+   using namespace @ModelName@_cxx_diagrams::fields;
 
-   if (std::abs(d) < 0.15) {
-      return (-0.33333333333333333333 +
-              0.25000000000000000000  * d -
-              0.20000000000000000000  * d * d +
-              0.16666666666666666667  * d * d * d -
-              0.14285714285714285714  * d * d * d * d +
-              0.12500000000000000000  * d * d * d * d * d -
-              0.11111111111111111111  * d * d * d * d * d * d +
-              0.10000000000000000000  * d * d * d * d * d * d * d -
-              0.090909090909090909091 * d * d * d * d * d * d * d * d +
-              0.083333333333333333333 * d * d * d * d * d * d * d * d * d -
-              0.076923076923076923077 * d * d * d * d * d * d * d * d * d * d);
-   }
+@EDMCalculation@
 
-   return 1.0 / (2.0 * d * d) * (3.0 - r - 2.0 * std::log(r) / d);
-}
-
-double OneLoopFunctionB(double r)
-{
-   if (is_zero(r))
-      return 1.0/2.0;
+   double val =
```

Comment:
I'd recommend to make this variable `const`.

Reply:
Done in 18b091268d2ffc1ad4a97c97771326d536681ea4.